src/pip/_internal/utils/retry.py

```diff
@@ -1,8 +1,15 @@
 import functools
 from time import perf_counter, sleep
-from typing import Callable, TypeVar
+from typing import TYPE_CHECKING, Callable, TypeVar
+
+if TYPE_CHECKING:
+    # TODO: import from typing once Python 3.10 is dropped.
+    from pip._vendor.typing_extensions import ParamSpec
```

Comment:
I think this should import directly from typing_extensions and then add it's to the mypy pre commit additional dependencies:

https://github.com/pypa/pip/blob/ab404769e0ffa3eca4aab9f0facd2fc07f7a360b/.pre-commit-config.yaml#L36

Reply:
Yeah, I need to fix the vendoring imports. `vendoring` is complaining because `pip._vendor.typing_extensions` does not exist anymore. I'll do that once Rich accepts my PR and cuts a new release.